### DISCUSSION TITLE
ZEN-garden on Euler

### DISCUSSION THREAD

**Discussion by jafluri** [ZEN-universe/ZEN-garden#186] [Installation & Setup] [answered]
# Prerequisites

**On Windows**
1.  Download mobaxterm.
2. Access the cluster via mobaxterm:
    - In mobaxterm, click "Sessions" on the top left. 
    - Click on the session type "SSH".
    - Enter "euler.ethz.ch" in the pop-up window next to "remote host" and click "ok".
    - Log in by entering you mail identifier, e.g. "mmuster" and your correspoding LDAP password.
    - Navigate to your home directory, e.g. "/cluster/home/mmuster". You can change directories using the `cd` command, e.g. `cd /cluster/home/mmuster`

**On MacOS**
1.  See below

# Installing ZEN-garden and the virtual environment

To run ZEN-Garden on Euler, you will need to create a virtual environment and load the necessary modules. 

## Download files

**When cloning a repository**
Corresponds to the local cloning described [here](https://zen-garden.readthedocs.io/en/latest/files/developer_guide/installation.html)

To set everything up, clone the repository on Euler:

1. Clone the github repository using `git clone https://github.com/ZEN-universe/ZEN-garden.git`. Note that you have to use the `https` and not the `ssh` url. If you want to clone a specific branch `git clone -b branchname https://github.com/ZEN-universe/ZEN-garden.git`. **If Github asks you for an access token**, create access token on github by clicking on your profile thumbnail->settings->developer settings->personal access tokens->generate new token and then setting the name _**and the required ticks**_. In the command on mobaxterm, you're asked for your github name and the password, which is the long ssh-link from the generated access token on github. **Save the token somewhere, you will not be able to access it later!**

2. Go into the `euler` directory. This can again be done using the `cd` command for changing directories: `cd ZEN-garden/euler/`. The `euler' directory contains shell scripts for running and executing ZEN-garden on the Euler cluster. 

**When using the direct installation**

Corresponds to the direct installation with `pip` described [here](https://zen-garden.readthedocs.io/en/latest/files/quick_start/installation.html)

1. Download `create_env_euler_pip.sh` and `source_env_euler.sh` directly from the GitHub. These are located in the `~\euler\` folder. They are shell scripts which help in setting up the ZEN-garden.
2. Copy the two files to Euler (e.g., by dragging them into the desired folder in Mobaxterm).

## Create environments
You can create the virtual environment by executing either the `create_env_euler.sh` (if cloning the repository) or the `create_env_euler_pip.sh` (if doing direct installation). These two files are henceforth collectively referred to as `create_env_euler(_pip).sh`, in which case the "_pip" should be included or omitted depending on your type of installation.

 To create a virtual environment, use one of these two following commands:

- When cloning a repository:
```
./create_env_euler.sh
```
- When using the direct installation
```
./create_env_euler_pip.sh
```
If you get a `Permission denied` error, you need to execute the [following command](https://linuxconfig.org/understanding-chmod-the-difference-between-ux-and-x):

```
chmod u+x create_env_euler(_pip).sh
```
Note that you do not get a confirmation that the permissions have changed.

Creating the environment can take several minutes. After completion, you can source the environment along with the required modules by running 

```
source source_env_euler.sh
```

To verify that everything works, you can see if your virtual environment contains `ZEN-garden`, i.e., you should find `zen_garden` in the list when typing `pip list`.

You will have to source the `source_env_euler.sh` file every time you log into Euler. However, you can add the following line

```
source <path/to/your>/source_env_euler.sh
```

in your `$HOME/.bash_profile` file, such that the environment loads automatically when you log in.

## Load Gurobi

To use Gurobi with ZEN-garden, you need to first load gurobi into Euler. The following command loads gurobi:

```
module load gurobi/12.0.1
```

Gurobi must be reloaded every time you start a new session. Alternatively, the above command can be added to the  `$HOME/.bash_profile` to automatically load gurobi in each session. To see what versions of Gurobi are available, see the [Euler documentation](https://scicomp.ethz.ch/wiki/Euler_applications_and_libraries_ubuntu)

If you still recieve an error message `AssertionError: Solver gurobi not installed` when running the ZEN-garden you may need to install the `gurobipy` package. To install this package, first active your ZEN-garden environment (as described in the previous section). Then, execute the following command:

```
pip install gurobipy
```

**Comment by jafluri** [comment, marked as answer]
**FOR A MORE HANDS-ON SETUP, READ THE LAST TWO ANSWERS (INCLUDING HOW TO SET STUFF UP ON MAC AND WINDOWS)**

**Standard Configuration**

A good configuration to use on Euler is 8 cores, 2 GB of RAM per core in the 4 hour queue. This leads to a setup that is as powerful as a good laptop. To run this in an interactive job use:

```
srun -t 04:00:00 -n 8 --mem-per-cpu=2048 --pty bash
```

If you just want to sumbit a command and then logout you can run

```
sbatch -t 04:00:00 -n 8 --mem-per-cpu=2048 --wrap=command
```

The main difference between `srun` and `sbatch` is that `srun` lets you commit only a single command and additionally can create an interactive environment. This means that you actually run on the computing nodes of Euler, you have full control about what you are doing etc. However, even if you are not running `srun` interactively, it will still run inside the current terminal, so you should not logout before the job has finished if you do not use `screen` or other additional commands. On the other hand `sbatch` lets you command full bash scripts in a "fire and forget" manner. The command above is tweaked to allow the submission of a single command and does not require the submission of a bash file. The `sbatch` command enables you to logout of Euler and check on your job later. Here is a summary of the features, also check the examples below:

| Feature | `srun` | `sbatch` |
|---|---|---|
| Interactive | possible | impossible |
| Runs in current terminal | yes | no |
| Waiting | Terminal needs to be open, might be long | can logout and check later, might be long |
| Monitoring | Easy | Medium |
|Debugging | Medium | Not recommended |
 

**Some general rule**

In case you need more resources, you should not exceed 4 GB of memery per core, i.e. `--mem-per-cpu=4096`. You can ask for as many cores as you want, however, if you ask for more than 32, i.e. `-n 32` you will have a low priority and the cores might end up on different nodes. Cores on different nodes can't work in parallel if you do not run MPI commands. If you wand to make absolutely sure that all your cores are on the same node, use `--ntasks-per-node=<n>` where you set it to the same value as the `-n` flag. If you need more time, increase the `-t` to 24 hours, going beyond that is quite bad for your priority and may lead to long waiting times.

For more details on commands, please refer to Euler's [quick reference](https://scicomp.ethz.ch/wiki/LSF_to_Slurm_quick_reference).

**Minimizing Resources**

After a job has finished, you can check the resources used by the job using the command:

```
seff <job_id>
```

Here, `<job_id>` should be replaced by the ID of the job (the 8 digit number located in the slurm output file name).  This will show you how much memory and CPU time was used by the job. Ideally, try to schedule jobs in such a way that the memory efficiency is at least 50%. Using too many resources costs the lab extra money and reduces your priority in the Euler queue. 


**Examples**

For both of the following examples you need to make sure that you have the necessary data on Euler in the `data` directory of the cloned repo, as well as a proper `config.py` file. Additionally, you should have created and sourced the virtual environment as described in the main post of the discussion above,

**Example 1**

To run the pipeline with `srun`, we start by using the command from above to create an interactive session, the output will look something like this:

```
(zen_garden_env) [jafluri@eu-login-06 data]$ srun -t 04:00:00 -n 8 --mem-per-cpu=2048 --pty bash
srun: job 4334654 queued and waiting for resources
srun: job 4334654 has been allocated resources
(zen_garden_env) [jafluri@eu-a2p-380 data]$
```

Note that the virtual environment was already loaded prior to the first line, theoretically one could do this also after the jobs starts. The first line starts the interactive session, Euler will then try to allocate the resources you asked for (i.e. number of cores, memory etc.), this is indicated by the line `queued and waiting for resources`. Note that waiting for resources can take multiple hours if you ask for a lot of cores and the cluster is busy. However, usually it takes only a couple of minutes. Once the resources have been allocated (`has been allocated resources`) you return to the normal shell environment. The only thing that changed is the remote, from `jafluri@eu-login-06`, indicating a Euler login note to `jafluri@eu-a2p-380` indicating a Euler compute note. Now you can use the package however you want or run completely different jobs, e.g.

```
(zen_garden_env) [jafluri@eu-a2p-380 data]$ python -m zen_garden
Running ZEN-Garden version: 0.1

--- Add elements to model---
...
```

The job will end after the time runs out or you run the exit command

```
(zen_garden_env) [jafluri@eu-a2p-380 data]$ exit
exit
(zen_garden_env) [jafluri@eu-login-06 data]$
```

Notice how the remote changed again from the compute node to the login node.

**Example 2**

To run the pipeline with `sbatch` you need to make sure that all necessary modules are loaded, i.e. the virtual environment with the source file. Then you simply submit a single command you want to run, e.g. to run the module:

```
(zen_garden_env) [jafluri@eu-login-06 data]$ sbatch -t 04:00:00 -n 8 --mem-per-cpu=2048 --wrap="python -m zen_garden"
Submitted batch job 4335397
```

This will return a jobid, if you submit multiple jobs, you can distinguish them by the id. Euler has now queued your command and will run it when the resources are available. You can check on your jobs with the `squeue -u $USER` command (`$USER` is your username which you can also use directly, e.g. `squeue -u jafluri`):

```
(zen_garden_env) [jafluri@eu-login-06 data]$ squeue -u $USER
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
           4335397 normal.4h     wrap  jafluri PD       0:00      1 (Priority)
```

We can see that the time is still 0:00 and the status (`ST`) is pending (`PD`), meaning that the job did not start yet. Theoretically, you could now logout and check later if the job started or finished. Once the job starts the output will look like this:

```
(zen_garden_env) [jafluri@eu-login-06 data]$ squeue -u $USER
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
           4335397 normal.4h     wrap  jafluri  R       0:12      1 eu-a2p-431
```

If you forgot what you submitted while the job is running you can list all the details (a lot) with the `scontrol` command and the job ID, e.g.

```
(zen_garden_env) [jafluri@eu-login-06 data]$ scontrol show jobid -dd 4335397
JobId=4335397 JobName=wrap
   UserId=jafluri(75756) GroupId=jafluri-group(158973) MCS_label=N/A
   Priority=1489 Nice=0 Account=normal/es_eawag QOS=es_eawag/normal
   JobState=RUNNING Reason=None Dependency=(null)
   Requeue=0 Restarts=0 BatchFlag=1 Reboot=0 ExitCode=0:0
   DerivedExitCode=0:0
   RunTime=00:00:09 TimeLimit=04:00:00 TimeMin=N/A
   SubmitTime=2022-12-15T10:14:02 EligibleTime=2022-12-15T10:14:02
   AccrueTime=2022-12-15T10:14:02
   StartTime=2022-12-15T10:14:58 EndTime=2022-12-15T14:14:58 Deadline=N/A
   SuspendTime=None SecsPreSuspend=0 LastSchedEval=2022-12-15T10:14:58 Scheduler=Backfill
   Partition=normal.4h AllocNode:Sid=eu-login-06:2954
   ReqNodeList=(null) ExcNodeList=(null)
   NodeList=eu-a2p-431
   BatchHost=eu-a2p-431
   NumNodes=1 NumCPUs=8 NumTasks=8 CPUs/Task=1 ReqB:S:C:T=0:0:*:*
   TRES=cpu=8,mem=16G,node=1,billing=69468
   Socks/Node=* NtasksPerN:B:S:C=0:0:*:1 CoreSpec=*
   JOB_GRES=(null)
     Nodes=eu-a2p-431 CPU_IDs=20,52,62-63,71-72,74-75 Mem=16384 GRES=
   MinCPUsNode=1 MinMemoryCPU=2G MinTmpDiskNode=0
   Features=(null) DelayBoot=00:00:00
   OverSubscribe=OK Contiguous=0 Licenses=(null) Network=(null)
   Command=(null)
   WorkDir=/cluster/scratch/jafluri/ZEN-garden/data
   StdErr=/cluster/scratch/jafluri/ZEN-garden/data/slurm-4335397.out
   StdIn=/dev/null
   StdOut=/cluster/scratch/jafluri/ZEN-garden/data/slurm-4335397.out
   Power=
```
This prints out all the details like, setup, paths. times, etc and is usually overkill. 

While the job is running, the output of the job is channeled per default to a `slurm-<jobid>.out` file in the directory that you submitted your job. This can be overwritten when you submit the job to a file that you specify. You can also see it from the output of the `scontrol` command in the `StdOut` line. You can monitor the job by looking into that file or check what happend after the run.

**Reply by mechen-gess** [reply]
is it `sbach` or `sbatch`?

**Reply by jafluri** [reply]
Sorry, was a typo, it's `sbatch`

**Comment by jjjcat7** [comment]
If you see this error when running on Euler "No module named zen_garden.__main__; 'zen_garden' is a package and cannot be directly executed" 

This is because of the setup of Python 3.9.9 on Euler. If you start seeing this error message, do not run Zen-garden under the data folder, but rather under the general ZEN-garden folder, which could solve this issue. Remember to specify the path to your config and dataset when needed

**Comment by johburger** [comment]
### Terminal and file transfer solution on MacOS:

If you're on a Mac, the whole process on Euler works exactly the same. However, mobaxterm is not available on Mac and you need something else as a terminal and for your file transfer. In the following, several options are described. It is recommended to first read the full article and then pick the solution which you find most suitable to your needs/level of expertise/coding preferences.

**Terminal:**
MacOS has a built in UNIX terminal (this app: <img width="35" alt="Screenshot 2023-12-12 at 14 06 46" src="https://github.com/ZEN-universe/ZEN-garden/assets/63449992/4c9e9904-2b95-4e4c-8fb8-2c55302164f5">) which you can use to access Euler. In order to do so, simply enter `ssh username@euler.ethz.ch` in your terminal and after pressing Enter insert your LDAP password. Afterwards you can follow the steps above.

**File transer:**
One option for transferring files is the `scp` protocol, which you can use inside the terminal. Important is that you use your local machine Terminal, i.e., not the Terminal window which is logged into Euler. On Mac you can simply open another Terminal tab with `⌘+T`. A simple description of the process can be found [here](https://kb.iu.edu/d/agye).

**However, this becomes quite cumbersome if you frequently transfer large quantities of files.** Therefore, other options for file transfer are available. 
[*FileZilla*](https://filezilla-project.org) is a free and open source program that works on Mac which can transfer files from your local machine to Euler. However, I have no first-hand experience with it. 
[*CyberDuck*](https://cyberduck.io) is another free FTP client for Mac. You can open a session with Euler if you enter `euler.ethz.ch` as the server and then your username and LDAP password. CyberDuck lets you store this server an the credentials so that you can reconnect quickly. 

**One program to do it all:**
Another solution is a commercial tool called [*Termius*](https://www.termius.com). If you register for the [GitHub Global Campus](https://education.github.com/students), you can get the Termius Pro version for free. Termius is an SSH and SFTP client in one program. It lets you quickly switch between a "Finder-like" file manager and a terminal. You can save Euler as a known host where your credentials are stored and then simply reconnect with a double-click. 
Personally, Termius is my preferred solution, even though it takes some more time to set everything up (GitHub Global campus, Termius license,...).

**Comment by jacob-mannhardt** [comment]
**This article is an extension of #186 and provides some additional hands-on tricks on how to use Euler most effectively.** 
The official [Euler Wiki](https://scicomp.ethz.ch/wiki/Main_Page) can solve most questions 


We want to touch upon the following topics:
1. `home` and `scratch`:
2. WinSCP (for Windows)
3. bash scripts

### `home` and `scratch`
Each user on Euler has access to two storages: [home and scratch](https://scicomp.ethz.ch/wiki/Storage_and_data_transfer)

Home is the default safe, long-term storage. That means everything that you upload to home will remain there. However, the disk quota is limited to 16 GB. You will not be able to store more than that. Optimizations that attempt to save data on a full disk will fail. 

Scratch, on the other hand, has a disk quota of 2.5 TB (per user!), but all files are deleted after two weeks, and no backups are made. The storage is not purged every two weeks but files that haven't been modified in two weeks will be deleted. If you update a dataset once a week, it will always remain on scratch. If you access scratch for the first time, your personal folder is not visible yet, but you can type it (see below) in a command line and find it.

A good workflow is to have source code on home and data/results on scratch. You can find both storages on Euler under:
`/cluster/home/username` or `$HOME`
`/cluster/scratch/username` or `$SCRATCH`
In ZEN-garden, we can point the framework at an arbitrary location of the config and the dataset (with the `--config` and `--dataset` option). This can be easily done in bash scripts (see below)

### WinSCP (for Windows)
The (in my opinion) easiest way to transfer files between Euler and your local Windows machine is [WinSCP](https://scicomp.ethz.ch/wiki/Storage_systems#Graphical_tools). After installing WinSCP, open the program. When asked to connect to a server, enter the following information: file protocol = `SFTP` and hostname = `euler.ethz.ch`. The user name is your usual ethz user name (without @ sign). You can save multiple sessions to quickly and efficiently jump between different folders on Euler.

To automatize the datatransfer, you can either [run WinSCP from a console](https://winscp.net/eng/docs/commandline) or use the [paramiko STFP client in Python](https://docs.paramiko.org/en/2.4/api/sftp.html).
### bash scripts
Once you have set up Euler, you might want to run multiple models with the same or similar settings. For this, we recommend bash scripts. A suitable location for the bash script is in the data folder of ZEN-garden on home. You should save the script with a `.sh` file ending. If you get a permission error when executing the script, type `chmod u+x <filename>` to set the [appropriate permissions](https://cets.seas.upenn.edu/answers/chmod.html). Note that you do not get a confirmation that the permissions have been changed.
One setup that I have been using is the following
```
#!/bin/bash

# define models
declare -a models=("model_1" "model_2" "model_3")

# current working directory
cwd="/cluster/scratch/username/folder_with_datasets/"

today=$(date +"%d%m")
config="$cwd/config.py"
# define time and mem options
t="4:00:00"
mem=512
n=48
# loop over models
echo "Run for models ${models[*]}"
for mod in "${models[@]}"
do
  # define dataset
  dataset="$cwd/$mod"
  sbatch --ntasks=$n --mem-per-cpu=$mem --time=$t --ntasks-per-node=$n  --output="$mod-$today-%j.out" --job-name="$mod-$today" --mail-type=END,FAIL --array=0-3 --wrap="python -m zen_garden --config=$config --dataset=$dataset"
done
```
We first declare a list of models (by their dataset name) that are found in the dataset folder, for example `/cluster/scratch/username/folder_with_datasets/` on scratch (see above). We then point at the config (`config="$cwd/config.py"`) and set the requested resources (`t`, `mem`, and `n`). We can loop over the models to start a job for each (`for mod in "${models[@]}"`) and set the dataset path (`dataset="$cwd/$mod"`). The most important line is the `sbatch` line which submits a job to euler. Besides the resource request and the wrap that tells Euler what to run (= ZEN-garden), there are the following considerations:
- When using the scenario analysis tool in ZEN-garden, we can parallelize the runs on Euler. For this to work, we must set the `--array` option. In the example above, we run 4 scenarios (0-3). Remember that the reference scenario is a scenario as well and must be included when calculating the number of jobs. 
- You can change the name of the job in the out file (`--output`) and in the `squeue` list (`--job_name`). This makes it easier to find the correct model output file later. `%j` is the unique job ID.
- By default, Euler distributes runs on various nodes. However, they don't communicate, so all resources should be allocated to the same node, which can be achieved by setting `--ntasks-per-node=$n`. Otherwise, Gurobi for example cannot parallelize the optimization. `n` must be smaller or equal to 128
